Sca rules

rzetelskik opened the pull request.

zimnx commented on pkg/recommender/metrics/prometheus_provider.go:
```diff
@@ -60,40 +65,63 @@ func discover(ctx context.Context, c client.Client, metricsSelector map[string]s
+	status := true
+	values := resultMatrix[0].Values //TODO check all results instead of just the first one???
+	for i := range values {
+		status = status && (values[i].Value != 0)
```
you can early exit when status becomes false

zimnx commented on pkg/recommender/metrics/provider.go:
```diff
@@ -2,15 +2,17 @@ package metrics
 
 type Provider interface {
-	FetchMetric(ctx context.Context, metric string, labels map[string]string) (float64, error)
+	Query(ctx context.Context, expression string) (bool, error)
```
It's a good practice to write short description of exposed functions in comment. Especially for exposed interfaces.

zimnx commented on pkg/recommender/recommender.go:
```diff
@@ -36,97 +41,243 @@ func New(ctx context.Context, c client.Client, logger log.Logger, metricsSelecto
+func isScyllaClusterReady(sc *scyllav1.ScyllaCluster) bool {
+	for _, rack := range sc.Spec.Datacenter.Racks {
+		rackStatus, found := sc.Status.Racks[rack.Name]
+		if !found || rackStatus.Members != rackStatus.ReadyMembers {
```
rackStatus.Members -> rack.Spec.Members

zimnx commented on pkg/recommender/recommender.go:
```diff
@@ -36,97 +41,243 @@ func New(ctx context.Context, c client.Client, logger log.Logger, metricsSelecto
+	for _, datacenterScalingPolicy := range scalingPolicy.Datacenters {
+		datacenter := sc.Spec.Datacenter
+		if datacenterScalingPolicy.Name != datacenter.Name {
+			return nil, errors.New(fmt.Sprintf("datacenter \"%s\" not found", datacenterScalingPolicy.Name))
```
nit: errors.Errorf()

zimnx commented on pkg/recommender/recommender.go:
```diff
@@ -36,97 +41,243 @@ func New(ctx context.Context, c client.Client, logger log.Logger, metricsSelecto
+			members = calculateMembers(rack.Members, min, max, rule.ScalingFactor)
+		} else {
+			if rack.Resources.Requests == nil || rack.Resources.Requests.Cpu() == nil {
+				return nil, errors.New("cpu requests undefined")
```
include info which rack doesn't have this info

rzetelskik replied:
the error is wrapped higher in the call stack (errors.Wrapf(err, "rack \"%s\"", rack.Name))

zimnx reviewed (commented):
Comments inline, please also add unit tests

rzetelskik committed:
Update vendor and go.sum

zimnx commented on pkg/recommender/test/recommender_test.go:
```diff
@@ -0,0 +1,562 @@
+package test
```
You don't have to create separate directory for test package. You can move these files to pkg/recommender and use `recommender_test` as package name

zimnx replied:
Usually package code and tests are shipped wihin the same directory. So for util package, you have util tests. Here you test utils as a part of recommender tests.

krzyssota replied:
Separated into recommender tests and prometheus provider tests.

zimnx commented on pkg/recommender/metrics/provider.go:
```diff
@@ -2,15 +2,31 @@ package metrics
+	// i.e. whether the expression was true in each data point in the given time range.
+	RangedQuery(ctx context.Context, expression string, duration time.Duration, argStep *time.Duration) (bool, error)
+
+	SetApi(api v1.API)
```
You're binding generic interface with concrete implementation. Why would one have to implement accepting prometheus/v1 for different metrics backend?

krzyssota replied:
Got rid of it.

zimnx commented on pkg/recommender/metrics/test_util.go:
```diff
@@ -0,0 +1,21 @@
+	"time"
+)
+
+func GetEmptyPrometheusProvider(logger log.Logger) Provider {
```
do not return it as interface, but as a pointer to concrete implementation.

krzyssota replied:
Solved it changing PrometheusProvider constructor.

zimnx commented on pkg/recommender/metrics/test_util.go:
```diff
@@ -0,0 +1,21 @@
+	}
+}
+
+func (p *prometheusProvider) SetApi(api v1.API) {
```
Instead requring this function in generic Provider interface, you can export it just for testing purposes.  
To have it available only in tests, you declare it as a part `metrics` package in `xxx_test.go` file (export_test.go for example). Then all `yyy_test.go` files within the same directory see this function as `metric.SetApi`, while non-test code cannot access it.

krzyssota replied:
Solved it changing PrometheusProvider constructor.

zimnx commented on pkg/recommender/test/recommender_test.go:
```diff
@@ -0,0 +1,562 @@
+		Level: atom,
+	})
+	r := recommender.GetEmptyRecommender(logger)
+	r.SetMetricsProviderFakeAPI(metrics.MockApi{})
```
Mock should be injected as a constructor parameter. Constructor should accept interface type, and your test should be able to supply such.

krzyssota replied:
Mock is injected as an (interface typed) constructor parameter.

`func NewPrometheusProvider(api v1.API, logger log.Logger, defaultStep time.Duration) (Provider, error) {
	return &prometheusProvider{
		provider: provider{
			logger:      logger,
			defaultStep: defaultStep,
		},
		api: api,
	}, nil
}`

zimnx commented on pkg/recommender/test/recommender_test.go:
```diff
@@ -0,0 +1,562 @@
+		errorExpected  bool
+	}{
+		{
+			name:           "Simple case",
```
Just by reading test description I don't understand what is being tested, this should be improved.

zimnx commented on pkg/recommender/recommender.go:
```diff
@@ -6,24 +6,31 @@ import (
-	}
+type Recommender interface {
+	RunOnce(ctx context.Context) error
+	/*SetMetricsProviderFakeAPI(api metrics.MockApi)
```
remove dead code

krzyssota replied:
Ok.

zimnx commented on pkg/recommender/recommender.go:
```diff
@@ -36,97 +43,225 @@ func New(ctx context.Context, c client.Client, logger log.Logger, metricsSelecto
-					r.logger.Error(ctx, "rack not found", "rack", rackScalingPolicy.Name)
-					continue
-				}
+func (r *recommender) ExportedQueryOnMetricsProvider(ctx context.Context, expression string) (bool, error) {
```
Changing application code just for testing purposes is bad practice. You should test packages as a blackbox, observing what is happening outside when exported function is called without having a need of calling private functions. You can mock/fake client.Client as well as metrics.Provider.

krzyssota replied:
"You should test packages as a blackbox, observing what is happening outside when exported function is called without having a need of calling private functions."
I tought unit test should cover single functions, some of which are private or are methods of interface private components (like Query of metricsProvider). Should I avoid exporting any function I do not have direct access to, i.e. writing only one big test for RunOnce (main function of recommender)?

krzyssota replied:
Used fake client. Application code is not changed, calling only exported functions.

zimnx commented on pkg/recommender/metrics/mock_prometheus_api.go:
```diff
@@ -0,0 +1,76 @@
+
+type MockApi struct{}
+
+var Q = func(query string, ts time.Time) (model.Value, v1.Warnings, error) {
```
Why this isn't a field of MockApi?

krzyssota replied:
Changed it.
`type mockApi struct {
	q  func(query string, ts time.Time) (model.Value, v1.Warnings, error)
	qr func(query string, r v1.Range) (model.Value, v1.Warnings, error)
}`

zimnx commented on pkg/recommender/metrics/mock_prometheus_api.go:
```diff
@@ -0,0 +1,76 @@
+	"time"
+)
+
+type MockApi struct{}
```
don't export it in application code. This should be moved to `prometheus_provider_mock_test.go` to be invisible outside of tests.

krzyssota replied:
Changed it.
`type mockApi struct {
	q  func(query string, ts time.Time) (model.Value, v1.Warnings, error)
	qr func(query string, r v1.Range) (model.Value, v1.Warnings, error)
}`

zimnx commented on pkg/recommender/test/recommender_test.go:
```diff
@@ -0,0 +1,562 @@
+			rack: getRackSpec(rackName, baseMembers, baseCpu, baseCpu, memory, memory),
+			scalingPolicy: getScalingPolicy(rackName,
+				[]v1alpha1.ScalingRule{
+					*getScalingRule(ruleName, 1, queryWillReturnTrue, nil, nil, v1alpha1.ScalingModeHorizontal, 2.0),
```
give a name to these magic numbers

krzyssota replied:
Ok. Added:
       const (
		rackName          = "rack_name"
		ruleName          = "rule_name"
		baseMembers       = 3
		baseCpu           = "5"
		higherCpu = "40"
		memory            = "1Gi"
		minAllowedMembers = 1
		maxAllowedMembers = 100
		priority1 = 1
		priority2 = 2
		priority3 = 3
		factor2 = 2.0
		factor4 = 4.0
		factor6 = 6.0
	)


	var (
		minAllowedCpu = resource.MustParse("1")
		maxAllowedCpu = resource.MustParse("100")
		duration5 = util.DurationPtr(5)
		duration10 = util.DurationPtr(10)
	)
	

and, if test results in scaling cpu, used it as follows:
	`corev1.ResourceCPU:    resource.MustParse(stringMulFloat64(baseCpu, factor4)),`

zimnx reviewed (changes requested):
go.sum change shouldn't be part of unit test commit
Tests should treat application as a blackbox, please refactor them to follow this appraoch of testing. 
Reorganization of code is needed.

rzetelskik committed:
SCA API rules implementation

- Implemented rules in SCA API
- Added vertical scaling support with options for requests and limits
- Changed recommender's logic so that it does not return recommendations if none of the rules. Added error propagation.
- Added timestamp to SCA's status
- Added update status to SCA's status
- Added ranged queries with resolution
- Move executables to bin directory

krzyssota committed:
Add Recommender unit tests

Co-authored-by: Kacper Rzetelski <rzetelski.k@gmail.com>

- Implemented recommender unit tests.
- Fixed computeCPU.
- Fixed rule priorities bug.
- Separate prometheus client creation from prometheus_provider constructor.
- Improve unit tests' names.
- Separate prometheus_provider tests from recommender tests.
- Test main exported function from recommender (RunOnce) using fake client.
- Client and custom mock_api.
- Fix recommender constructor return type.
- Change tests' filenames.
- Aadd const port.

zimnx commented on pkg/recommender/metrics/prometheus_provider.go:
```diff
@@ -18,38 +17,39 @@ import (
 	addr := (&url.URL{
 		Scheme: "http",
-		Host:   fmt.Sprintf("%s.%s.svc.cluster.local", svc.Name, svc.Namespace),
+		Host:   fmt.Sprintf("%s.%s.svc.cluster.local:9090", svc.Name, svc.Namespace),
```
9090 could be a const

rzetelskik replied:
added const svcPort

zimnx commented on pkg/recommender/test_util.go:
```diff
@@ -0,0 +1,246 @@
+package recommender
```
rename to util_test to make this code visible only in tests. If it's used in other packages, extract it to package.

rzetelskik replied:
renamed to util_test.go

zimnx commented on pkg/recommender/metrics/mock_prometheus_api/mock_prometheus_api.go:
```diff
@@ -0,0 +1,99 @@
+package mock_prometheus_api
```
just `mock` is enough

rzetelskik replied:
changed to mock

zimnx commented:
Much better, thanks, left few nits.

zimnx reviewed (approved):
LGTM, thanks
